File: prolog/words/src/numerical.pl
fibonacci(N, Result) :-
    fibonacci_iter(N, 0, 1, Result), !.

fibonacci_iter(0, Acc2, _, Acc2).
fibonacci_iter(N, _, _, Result) :-
    N < 0,
    Result is -1.
fibonacci_iter(N, Acc2, Acc1, Result) :-
    N =< 2,
    Result is Acc1 + Acc2.
fibonacci_iter(N, Acc2, Acc1, Result) :-
    N1 is N - 1,
    fibonacci_iter(N1, Acc1, Acc1 + Acc2, Result), !.

factorial(N, Result) :-
    factorial_iter(N, 1, Result), !.

factorial_iter(N, _, Result) :-
    N < 0,
    Result is -1.
factorial_iter(N, Acc, Result) :-
    N =< 1,
    Result is Acc.
factorial_iter(N, Acc, Result) :-
    N1 is N - 1,
    factorial_iter(N1, Acc * N, Result), !.

gcd(A, 0, A).
gcd(A, B, Result) :-
    B > 0,
    R1 is A mod B,
    gcd(B, R1, Result), !.

lcm(A, B, Result) :-
    gcd(A, B, GCD),
    Result is (A / GCD) * B.

File: _experimental/_c++/words/src/numerical.cpp
#include "../include/numerical.h"

// Implementation of the methods
int Numerical::fibonacci(int n) {
    if (n < 0) return -1;
    else if (n < 2) return n;
    int acc2 = 0, acc1 = 1, aux = 0;
    while(n-- > 2) {
        aux = acc1 + acc2;
        acc2 = acc1;
        acc1 = aux;
    }
    
    return acc1 + acc2;
}

long Numerical::factorial(int n) {
    if (n < 0) return -1;
    long acc = 1;
    while(n > 1) {
        acc *= n--;
    }
    
    return acc;
}

int Numerical::gcd(int a, int b) {
    int aux = 0;
    while(b != 0) {
        aux = a % b;
        a = b;
        b = aux;
    }
    return a;
}

int Numerical::lcm(int a, int b) {
    if (a == 0 || b == 0) return 0;
    return (a / gcd(a, b)) * b;
}